gateway/faux-mgs/src/usart.rs

```diff
@@ -0,0 +1,317 @@
+        termios::tcsetattr(stdout, termios::TCSAFLUSH, &ios)
+            .with_context(|| "failed to set termios on stdout")?;
+        Some(UnrawTermiosGuard { stdout, ios: orig_ios })
+    } else {
+        None
+    };
+
+    let socket = Arc::new(socket);
+    let rx_sp_msg = {
+        let (tx, rx) = crossbeam_channel::bounded(16);
+        let log = log.clone();
+        let socket = Arc::clone(&socket);
+        thread::spawn(move || recv_task(log, socket, tx));
+        rx
+    };
+    let rx_stdin = {
+        let (tx, rx) = crossbeam_channel::bounded(16);
+        let log = log.clone();
+        thread::spawn(move || stdin_task(log, raw, tx));
+        rx
+    };
+
+    let mut packetizer =
+        SerialConsolePacketizer::new(SpComponent::try_from("sp3").unwrap());
+    loop {
+        select! {
+            recv(rx_sp_msg) -> msg => {
+                // recv_task() runs as long as we're holding `rx_sp_msg`, so we
+                // can unwrap here.
+                handle_sp_message(&log, msg.unwrap(), None);
+            }
+
+            recv(rx_stdin) -> result => {
+                let data = match result {
+                    Ok(line) => line,
+                    Err(_) => break,
+                };
+
+                for chunk in packetizer.packetize(&data) {
+                    let request_id = crate::send_request(
+                        &log,
+                        &socket,
+                        sp,
+                        RequestKind::SerialConsoleWrite(chunk),
+                    )?;
+
+                    // Don't read any more from our stdin until we've gotten
+                    // acks that this line as been received.
+                    for msg in rx_sp_msg.iter() {
+                        if handle_sp_message(&log, msg, Some(request_id)) {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+struct UnrawTermiosGuard {
+    stdout: i32,
+    ios: Termios,
+}
+
+impl Drop for UnrawTermiosGuard {
+    fn drop(&mut self) {
+        termios::tcsetattr(self.stdout, termios::TCSAFLUSH, &self.ios).unwrap();
+    }
+}
+
+/// If `request_id` is `Some(n)`, returns `true` if `msg` contains a serial
+/// console ack of that request id. Returns `false` in all other cases.
+fn handle_sp_message(
+    log: &Logger,
+    msg: SpMessage,
+    request_id: Option<u32>,
+) -> bool {
+    match msg.kind {
+        SpMessageKind::Response { request_id: response_id, result } => {
+            match result {
+                Ok(ResponseKind::SerialConsoleWriteAck) => {
+                    if Some(response_id) == request_id {
+                        debug!(log, "received expected ack");
+                        return true;
+                    } else {
+                        debug!(log, "received unexpected ack");
+                    }
+                }
+                Ok(other) => {
```

Comment:
This seems to imply that we are only handling responses from the SP. Does the SP ever send async messages?

Reply:
It definitely can, and MGS proper has a lot more machinery to handle that. `faux-mgs` may eventually need to handle this better, but for now the only async messages the SP can send contain serial console data, which it won't send unless an MGS instance has already attached by sending it data first.

Reply:
Thanks. Makes sense.